File: ExtremeParser/Parsers/PawnDocParser.vb
Public Class PawnDocParser

#Region "Public Properties"
    Public ReadOnly Property Summary As String
        Get
            Return p_summary
        End Get
    End Property
    Public ReadOnly Property Parameters As List(Of KeyValuePair(Of String, String))
        Get
            Return p_params
        End Get
    End Property
    Public ReadOnly Property Returns As String
        Get
            Return p_returns
        End Get
    End Property
    Public ReadOnly Property Remarks As String
        Get
            Return p_remarks
        End Get
    End Property
#End Region

    Private p_summary As String
    Private p_params As New List(Of KeyValuePair(Of String, String))
    Private p_returns As String
    Private p_remarks As String

    Public Sub New(text As String)
        'Prepare the text for parsing.
        Dim lines As String() = text.Split(vbCrLf.ToCharArray) 'Cut it into multiple lines.

        'Check for posible errors.
        If text.Contains("<summary>") And Not text.Contains("</summary>") Then Throw New ParserException("The summary closing tag is missing.", "")
        If text.Contains("<returns>") And Not text.Contains("</returns>") Then Throw New ParserException("The returns closing tag is missing.", "")
        If text.Contains("<remarks>") And Not text.Contains("</remarks>") Then Throw New ParserException("The remarks closing tag is missing.", "")

        'Configs.
        Dim isInSummary As Boolean = False : Dim currentSummary As String = Nothing
        Dim isInAParam As Boolean = False : Dim currentParamName As String = Nothing : Dim currentParam As String = Nothing : Dim currentParamDesc As String = Nothing
        Dim isInReturn As Boolean = False : Dim currentReturn As String = Nothing
        Dim isInRemarks As Boolean = False : Dim currentRemark As String = Nothing

        For Each line As String In lines
            line = line.Trim 'Remove the whitespace.

            If isInSummary Then
                currentSummary = currentSummary + IIf(line = "", "", line + vbCrLf) 'Add the old one.
                If currentSummary.EndsWith("</summary>" + vbCrLf) Then
                    currentSummary = currentSummary.Replace("</summary>" + vbCrLf, "")
                    p_summary = currentSummary.Trim()
                    isInSummary = False
                End If
                Continue For
            ElseIf isInAParam Then
                currentParam = currentParam + IIf(line = "", "", line + vbCrLf) 'Add the old one.
                If currentParam.EndsWith("</param>" + vbCrLf) Then
                    currentParam = currentParam.Replace("</param>" + vbCrLf, "")
                    currentParamDesc = currentParam.Trim()
                    isInAParam = False

                    'Create the key
                    Parameters.Add(New KeyValuePair(Of String, String)(currentParamName, currentParamDesc))
                End If
                Continue For
            ElseIf isInReturn Then
                currentReturn = currentReturn + IIf(line = "", "", line + vbCrLf) 'Add the old one.
                If currentReturn.EndsWith("</returns>" + vbCrLf) Then
                    currentReturn = currentReturn.Replace("</returns>" + vbCrLf, "")
                    p_returns = currentReturn.Trim()
                    isInReturn = False
                End If
                Continue For
            ElseIf isInRemarks Then
                currentRemark = currentRemark + IIf(line = "", "", line + vbCrLf) 'Add the old one.
                If currentRemark.EndsWith("</remarks>" + vbCrLf) Then
                    currentRemark = currentRemark.Replace("</remarks>" + vbCrLf, "")
                    p_remarks = currentRemark.Trim()
                    isInRemarks = False
                End If
                Continue For
            End If

            If line.StartsWith("<summary>") Then
                line = line.Replace("<summary>", "")
                currentSummary = line + vbCrLf

                If line.EndsWith("</summary>") Then
                    line = line.Replace("</summary>", "")
                    p_summary = line.Trim()
                Else
                    isInSummary = True
                End If
                Continue For
            ElseIf line.StartsWith("<param name=" + Chr(34)) Then
                line = line.Replace("<param name=" + Chr(34), "")
                currentParamName = line.Substring(0, line.IndexOf(Chr(34)))
                line = line.Remove(0, line.IndexOf(">") + 1)

                currentParam = line + vbCrLf
                If line.EndsWith("</param>") Then
                    line = line.Replace("</param>", "")
                    currentParamDesc = line.Trim()

                    'Create the key
                    Parameters.Add(New KeyValuePair(Of String, String)(currentParamName, currentParamDesc))
                Else
                    isInAParam = True
                End If
            ElseIf line.StartsWith("<returns>") Then
                line = line.Replace("<returns>", "")
                currentReturn = line + vbCrLf

                If line.EndsWith("</returns>") Then
                    line = line.Replace("</rturns>", "")
                    p_returns = line.Trim()
                Else
                    isInReturn = True
                End If
                Continue For
            ElseIf line.StartsWith("<remarks>") Then
                line = line.Replace("<remarks>", "")
                currentRemark = line + vbCrLf

                If line.EndsWith("</remarks>") Then
                    line = line.Replace("</remarks>", "")
                    p_remarks = line.Trim()
                Else
                    isInRemarks = True
                End If
                Continue For
            End If
        Next
    End Sub
End Class

File: ExtremeParser/Parsers/PawnDocParser.cs
using System;
using System.Collections.Generic;
using ExtremeParser.Exceptions;

namespace ExtremeParser.Parsers
{
    public class PawnDocParser
    {

        #region Public Properties

        public string Summary
        {
            get { return _pSummary; }
        }

        public List<KeyValuePair<string, string>> Parameters
        {
            get { return _pParams; }
        }

        public string Returns
        {
            get { return _pReturns; }
        }

        public string Remarks
        {
            get { return _pRemarks; }
        }

        #endregion

        private string _pSummary;
        private List<KeyValuePair<string, string>> _pParams = new List<KeyValuePair<string, string>>();
        private string _pReturns;
        private string _pRemarks;

        public PawnDocParser(string text)
        {
            //Prepare the text for parsing.
            string[] lines = text.Split('\r', '\n'); //Cut it into multiple lines.

            //Check for posible errors.
            if (text.Contains("<summary>") && !text.Contains("</summary>"))
            {
                throw (new ParserException("The summary closing tag is missing.", ""));
            }

            if (text.Contains("<returns>") && !text.Contains("</returns>"))
            {
                throw (new ParserException("The returns closing tag is missing.", ""));
            }

            if (text.Contains("<remarks>") && !text.Contains("</remarks>"))
            {
                throw (new ParserException("The remarks closing tag is missing.", ""));
            }

            //Configs.
            bool isInSummary = false;
            string currentSummary = null;
            bool isInAParam = false;
            string currentParamName = null;
            string currentParam = null;
            string currentParamDesc = null;
            bool isInReturn = false;
            string currentReturn = null;
            bool isInRemarks = false;
            string currentRemark = null;

            foreach (string tmpline in lines)
            {
                string line = tmpline;
                line = line.Trim(); //Remove the whitespace.

                if (isInSummary)
                {
                    currentSummary =
                        Convert.ToString(currentSummary +
                                                (string.IsNullOrEmpty(line) ? "" : line + "\r\n")); //Add the old one.
                    if (currentSummary.EndsWith(Convert.ToString("</summary>" + "\r\n")))
                    {
                        currentSummary = currentSummary.Replace(Convert.ToString("</summary>" + "\r\n"), "");
                        _pSummary = currentSummary.Trim();
                        isInSummary = false;
                    }

                    continue;
                }

                if (isInAParam)
                {
                    currentParam =
                        Convert.ToString(currentParam +
                                         (string.IsNullOrEmpty(line) ? "" : line + "\r\n")); //Add the old one.
                    if (currentParam.EndsWith(Convert.ToString("</param>" + "\r\n")))
                    {
                        currentParam = currentParam.Replace(Convert.ToString("</param>" + "\r\n"), "");
                        currentParamDesc = currentParam.Trim();
                        isInAParam = false;

                        //Create the key
                        Parameters.Add(new KeyValuePair<string, string>(currentParamName, currentParamDesc));
                    }

                    continue;
                }

                if (isInReturn)
                {
                    currentReturn =
                        Convert.ToString(currentReturn +
                                         (string.IsNullOrEmpty(line) ? "" : line + "\r\n")); //Add the old one.
                    if (currentReturn.EndsWith(Convert.ToString("</returns>" + "\r\n")))
                    {
                        currentReturn = currentReturn.Replace(Convert.ToString("</returns>" + "\r\n"), "");
                        _pReturns = currentReturn.Trim();
                        isInReturn = false;
                    }

                    continue;
                }

                if (isInRemarks)
                {
                    currentRemark =
                        Convert.ToString(currentRemark +
                                         (string.IsNullOrEmpty(line) ? "" : line + "\r\n")); //Add the old one.
                    if (currentRemark.EndsWith(Convert.ToString("</remarks>" + "\r\n")))
                    {
                        currentRemark = currentRemark.Replace(Convert.ToString("</remarks>" + "\r\n"), "");
                        _pRemarks = currentRemark.Trim();
                        isInRemarks = false;
                    }

                    continue;
                }

                if (line.StartsWith("<summary>"))
                {
                    line = line.Replace("<summary>", "");
                    currentSummary = line + "\r\n";

                    if (line.EndsWith("</summary>"))
                    {
                        line = line.Replace("</summary>", "");
                        _pSummary = line.Trim();
                    }
                    else
                    {
                        isInSummary = true;
                    }
                }
                else if (line.StartsWith("<param name=" + (char) 34))
                {
                    line = line.Replace("<param name=" + (char) 34, "");
                    currentParamName = line.Substring(0, line.IndexOf((char) 34));
                    line = line.Remove(0, line.IndexOf(">") + 1);

                    currentParam = line + "\r\n";
                    if (line.EndsWith("</param>"))
                    {
                        line = line.Replace("</param>", "");
                        currentParamDesc = line.Trim();

                        //Create the key
                        Parameters.Add(new KeyValuePair<string, string>(currentParamName, currentParamDesc));
                    }
                    else
                    {
                        isInAParam = true;
                    }
                }
                else if (line.StartsWith("<returns>"))
                {
                    line = line.Replace("<returns>", "");
                    currentReturn = line + "\r\n";

                    if (line.EndsWith("</returns>"))
                    {
                        line = line.Replace("</rturns>", "");
                        _pReturns = line.Trim();
                    }
                    else
                    {
                        isInReturn = true;
                    }
                }
                else if (line.StartsWith("<remarks>"))
                {
                    line = line.Replace("<remarks>", "");
                    currentRemark = line + "\r\n";

                    if (line.EndsWith("</remarks>"))
                    {
                        line = line.Replace("</remarks>", "");
                        _pRemarks = line.Trim();
                    }
                    else
                    {
                        isInRemarks = true;
                    }
                }
            }
        }
    }
}